### PR TITLE
Fix sNaN behavior of .minimum, .maximum, .minimumMagnitude and .maximumMagnitude.

### PR DESCRIPTION
**Pull request by stephentyrone**
Specifically, if either argument is a signaling NaN, the result should be a quiet NaN.  Previously, signaling NaN inputs were treated identically to quiet NaN inputs.

Some other changes folded in with this:
- Formally define which argument is returned if lhs == hrs or abs(lhs) == abs(rhs) in the case of the Magnitude versions.  This guarantees that the set {min,max} is identical to the set {lhs,rhs} so long as neither argument is NaN.  This is not only a nod to formalism; it's also a critical property for generic head-tail arithmetic to simulate wider types.
- Added test coverage for these four operations.

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

**Comment by stephentyrone** [comment]
@swift-ci please test.

**Comment by stephentyrone** [comment]
@gribozavr I would appreciate another set of eyes on this if you can spare a few minutes.

**Comment by stephentyrone** [comment]
OS X build failure is Foundation noise unrelated to this.

File: test/1_stdlib/FloatingPoint.swift.gyb [hunk 1]
```diff
@@ -470,6 +477,61 @@ FloatingPoint.test("${FloatSelf}/{Comparable,Hashable,Equatable}") {
...
+        expectTrue(max.isQuietNaN)
+        expectTrue(minMag.isQuietNaN)
+        expectTrue(maxMag.isQuietNaN)
+      }
+      //  If only one of lhs and rhs is NaN, the result of the min/max
+      //  operations is always the other value.  While contrary to all other
+      //  IEEE 754 basic operations, this property is critical to ensure
+      //  that clamping to a valid range behaves as expected.
+      else if lhs.isNaN && !rhs.isNaN {
+        print("lhs is \(lhs.bitPattern).  rhs is \(rhs.bitPattern).")
```
Comment (gribozavr) [inline test/1_stdlib/FloatingPoint.swift.gyb]: Debugging code?

Comment (stephentyrone) [inline test/1_stdlib/FloatingPoint.swift.gyb]: Yup, thanks.  Will push a fix shortly.

**Comment by gribozavr** [comment]
LGTM, that's how these operations signal an invalid operation exception.

**Pull request merged**